### PR TITLE
feat: implement PC opcode

### PR DESCRIPTION
**Pull request by lambda-0x**
## Pull Request type

Please check the type of change your PR introduces:

- [x] Feature

## What is the current behavior?

Resolves: #254 #255 

## What is the new behavior?

- implements `PC` opcode
- kakarot on cairo0 just increments the PC manually, which I don't think will be very useful
- i didn't find tests related PC specifically in geth.
- evm.codes also dont specify any examples.

## Does this introduce a breaking change?

- [x] No

### CODE REVIEW

File: crates/evm/src/tests/test_instructions/test_memory_operations.cairo [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    // When
+    ctx.exec_pc();
+
+    // Then
+    assert(ctx.stack.len() == 1, 'stack should have one element');
+    assert(ctx.stack.pop().unwrap() == 0, 'PC should be 0');
+}
+// TODO: Test that PC updates properly when instructions are executed
+
```
Comment (lambda-0x) [inline crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: i am not sure if it would be a good idea to add a test for each instruction manually, instead we can make sure that each instruction is updating the PC as expected. 

both kakarot in cairo 0 and geth doesn't seem it test PC opcode manually.

Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: we need to call `decode_and_execute` with proper context to actually update the PC, any ideas on how to test all the cases cleanly?

Comment (khaeljy) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: Maybe wait until issue #250 is implemented, but that wouldn't really be a unit test...

Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: i am not sure how that would be useful here?

Comment (khaeljy) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: You can execute JUMP to a given index, and then you can check whether the new PC has the right value.

Comment (khaeljy) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: Or simply use `ctx.program_counter = GIVEN_INDEX`

Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: what would that be testing though? 

the point of testing is that you have an `initial setup` -> `do some processing` -> `assert on the resulting` state.

If we set PC manually it won't test that running an instruction produces correct PC (maybe checking for correctness maybe out of scope of this PR but i just wanted to point that out)

Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: does this PR needs to check that all opcodes results in a proper PC value?

cc: @enitrat @Eikix

Comment (khaeljy) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: ah ok sorry, I didn't quite get your point.

Comment (ClementWalter) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: there is actually a test in kakarot 0, see `test__exec_pc__should_update_after_incrementing`

regarding what to test from a unit testing perspective:
- the `exec_pc` opcode is supposed to put whatever the PC value is in the Stack, so setting the PC to something and asserting that the stack got the same value is fine
- the `exec_pc` is not responsible for actually incrementing the PC, so we shall not test that PC is incremented properly in a PC unit test
- as you pointed out, `decode_and_execute` is responsible for incrementing the PC, and asserting this increment is done properly will be done in a `decode_and_execute` test

**Review by ClementWalter — CHANGES_REQUESTED** [review]
please just do the same test as in kakarot (see https://github.com/ClementWalter/kakarot/blob/b6619adb6092984f4c6b4fa512ebf07ec8d47299/tests/src/kakarot/instructions/test_memory_operations.cairo#L20).

ie put a value and test that PC returns it, for several values

File: crates/evm/src/tests/test_instructions/test_memory_operations.cairo [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    ExecutionContext, ExecutionContextTrait, BoxDynamicExecutionContextDestruct, CallContextTrait,
+};
+
+
+#[test]
+#[available_gas(20000000)]
+fn test_pc_basic() {
+    // Given
+    let mut ctx = setup_execution_context();
+
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_memory_operations.cairo:21]: can you add an inlined `set_pc` method to the 
 `ExecutionContext` struct implementation, and manually set this value in the test to assert that the value popped is not 0 (default value)

File: crates/evm/src/tests/test_instructions/test_memory_operations.cairo (hunk 1 above)
Comment (enitrat) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: agreed with clement

Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:30]: i thought because we where setting the value manually the test doesn't make much sense, but yeah when you think about that way it make sense. will add.

File: crates/evm/src/tests/test_instructions/test_memory_operations.cairo (hunk 2 above)
Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo:21]: done

File: crates/evm/src/tests/test_instructions/test_memory_operations.cairo [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    ctx.exec_pc();
+
+    // Then
+    assert(ctx.stack.len() == 1, 'stack should have one element');
+    assert(ctx.stack.pop().unwrap() == 9000, 'updating PC failed');
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_pc_gets_updated_properly_2() {
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_memory_operations.cairo:48]: only 1 test of this kind is enough

**Comment by ClementWalter** [comment]
> only 1 test of this kind is enough 

we dramatically miss a way to parametrize the tests, in cairo0 we had tests for PC in `range(1, 15)`

**Comment by enitrat** [comment]
> we dramatically miss a way to parametrize the tests, in cairo0 we had tests for PC in range(1, 15)

Agreed but here the logic is fairly simple and doesn't present any other edge cases than pc > range bound. I wouldn't duplicate test for no reason

File: crates/evm/src/tests/test_instructions/test_memory_operations.cairo (hunk 3 above)
Comment (lambda-0x) [reply crates/evm/src/tests/test_instructions/test_memory_operations.cairo]: thought so but because clement told to test few cases, i added it anyway.

**Review by enitrat — APPROVED** [review]
lgtm

**Review by ClementWalter — APPROVED** [review]
